log/scope.go

```diff
@@ -315,3 +343,149 @@ func (s *Scope) SetLogCallers(logCallers bool) {
+	copy(out.handlers, s.handlers)
+	return &out
+}
+
+// AddContext adds a key-value pair to s, which is treated as context information.
+func (s *Scope) AddContext(key string, value interface{}) *Scope {
+	mu.Lock()
+	defer mu.Unlock()
+	s.context[key] = value
+	s.contextKey = append(s.contextKey, key)
+	return s
+}
+
+// WithContext returns a copy of s with the given key-value appended to the context.
+func (s *Scope) WithContext(key string, value interface{}) *Scope {
+	out := s.Local()
+	out.AddContext(key, value)
+	return out
+}
+
+// ClearContextKey clears a context by key in s and returns the resulting Scope.
+// If the key is not found, returns s.
+func (s *Scope) ClearContextKey(key string) *Scope {
+	mu.Lock()
+	defer mu.Unlock()
+	delete(s.context, key)
+	s.contextKey = removeKey(s.contextKey, key)
+	return s
+}
+
+// ClearAllContext clears all context in s and returns the resulting Scope.
+func (s *Scope) ClearAllContext() *Scope {
+	mu.Lock()
+	defer mu.Unlock()
+	s.context = make(map[string]interface{})
+	s.contextKey = nil
+	return s
+}
+
+// AddLabel adds a key-value pair to the labels in s.
+func (s *Scope) AddLabel(key string, value interface{}) *Scope {
+	mu.Lock()
+	defer mu.Unlock()
+	s.structured[key] = value
+	s.structuredKey = append(s.structuredKey, key)
+	return s
+}
+
+// WithLabel makes a copy of s with the given key-value appended to its labels.
+func (s *Scope) WithLabel(key string, value interface{}) *Scope {
+	out := s.Local()
+	out.AddLabel(key, value)
+	return out
+}
+
+// ClearLabelKey cleans a label in s with the given key and returns the resulting Scope.
+// If the key is not found, returns s.
+func (s *Scope) ClearLabelKey(key string) *Scope {
+	mu.Lock()
+	defer mu.Unlock()
+	delete(s.structured, key)
+	s.structuredKey = removeKey(s.structuredKey, key)
+	return s
+}
+
+// ClearAllLabels clears all labels from s and returns the resulting Scope.
+func (s *Scope) ClearAllLabels() *Scope {
+	mu.Lock()
+	defer mu.Unlock()
+	s.structured = make(map[string]interface{})
+	s.structuredKey = nil
+	return s
+}
+
+// callHandlers calls all handlers registered to s.
+func (s *Scope) callHandlers(
+	severity Level,
+	scope *Scope,
+	ie *errdict.IstioErrorStruct,
+	msg string,
+	fields []zapcore.Field) {
+
+	mu.RLock()
+	scope.mu.RLock()
+	hs := append(defaultHandlers, scope.handlers...)
+	scope.mu.RUnlock()
+	mu.RUnlock()
+
+	for _, h := range hs {
+		h(severity, scope, ie, msg, fields)
+	}
+}
+
+// getErrorStruct returns (*IstioErrorStruct, 1) if it is the first argument in the list is an IstioErrorStruct ptr,
+// or (nil,0) otherwise. The second return value is the offset to the first non-IstioErrorStruct field.
+func getErrorStruct(fields ...interface{}) (*errdict.IstioErrorStruct, int) {
+	ief, ok := fields[0].([]interface{})
+	if !ok {
+		return nil, 0
+	}
+	ie, ok := ief[0].(*errdict.IstioErrorStruct)
+	if !ok {
+		return nil, 0
+	}
+	// Skip IstioErrorStruct, pass remaining fields on as before.
+	return ie, 1
```

Comment:
clarifying: to use the structured error, it *must* be the first argument in the `Debuga` (or similar) call?  Is the intent to make the API the same while adding the functionality, then?

Is adding a bunch of `DebugStructured()`, etc.,  that enforce the ordering directly problematic because of all the levels ?

Reply:
Right, exactly. An alternative would be to name the calls something different (InfoStructf, DebugStructf...). 
I'm ok with either approach. The second approach would be a brand new thing rather than trying for backward compatibility which could actually be a good thing.